### PR TITLE
graphics: Scene, hit regions, subtree bitmap cache and Tween for CView

### PR DESCRIPTION
**Pull request by ajeetgill**
`foam.graphics.CView` is a scene graph with paint and hit-test, and every app that wants a pannable canvas builds the camera, the tooltip and the redraw budget on top of it again. This PR adds those pieces once, as small classes beside `CView`, borrowing the three techniques Konva uses: hit regions, a bitmap cache per subtree, and a tween.

- **Scene** — root `CView` that owns the camera (`x`, `y`, `zoom`), the viewport size and ordered `SceneLayer`s; `zoomAt` zooms about a point, `hitAtView` walks layers topmost first so an overlay never swallows a click meant for the content under it.

- **HitRegion** — invisible rectangle a shape adds over one clickable sub-part; `findFirstChildAt` returns it and the owner reads `role`, so a shape with three hot spots needs no event system of its own.

- **TooltipCView** — one tooltip box in the overlay layer, sized from measured text, multi-line, never a hit target.

- **CView.cache() / uncache()** — render a subtree to an offscreen bitmap once and blit it until a property inside the subtree changes; re-renders when the effective scale drifts more than `CACHE_SCALE_TOLERANCE`, pads by `CACHE_PAD` so edge strokes survive, and falls back to live painting when no canvas is available (tests, server). The bitmap lives in plain fields rather than properties, since a property write is exactly the signal that invalidates the cache.

- **Tween** — one eased 0..1 progression over a duration with injectable scheduler and clock; the caller maps the value onto a stroke width or a camera position.

- **CViewTheme** and **TextUtil** — colour/font tokens with a loud magenta fallback for a missing token, and text truncation with an injectable measurer.

The demo at `src/foam/demos/scene/` (linked from the demos index) pans, zooms, shows tooltips and hit regions over a 600-box grid, and prints the paint cost with the content layer cached versus live.

Tests: `SceneTest`, `HitRegionTest`, `CViewCacheTest`, `CViewThemeTest`, `TextUtilTest`, `TweenTest` under `src/foam/graphics/test/` (48 assertions), run with `./build.sh --client-tests:SceneTest,HitRegionTest,CViewCacheTest,CViewThemeTest,TextUtilTest,TweenTest`.

Nothing existing changes behaviour: `CView.paint` takes the cached branch only after `cache()` is called. The Reflow graph in #5359 paints with plain `CView` and can pick up `Scene`, `TooltipCView` and `cache()` later if that helps; nothing here depends on it or blocks it. A first consumer, a grammar viewer built on these classes, follows in a separate PR.

### CODE REVIEW

File: src/foam/graphics/CView.js [hunk 1]
```diff
@@ -715,10 +818,15 @@ foam.CLASS({
           x.shadowBlur  = shadowBlur;
         }
 
-        x.save();
-          this.paintSelf(x);
-        x.restore();
-        this.paintChildren(x);
+        if ( this.cached_ ) {
+          // Re-render when stale or when the zoom drifted enough to blur; otherwise blit the bitmap.
+          var scaleNow = this.effectiveScale_(x) * this.scaleX;
```
Comment (abdelaziz-mahdy) [inline src/foam/graphics/CView.js]: [bug] A cached node with a scaleX other than 1 allocates four times the pixels it needs.

doTransform has already put this node's scaleX on the context by the time effectiveScale_ reads it
back, so multiplying by this.scaleX again squares it. The picture is right — drawImage rescales —
it is just memory and render time. Same expression in renderCache_.

Drop the `* this.scaleX`.

The tests do not catch it because the fake context has no getTransform, so effectiveScale_ returns 1.

File: src/foam/graphics/Scene.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+      name: 'measure',
+      documentation: 'function(text, font) -> width in px. Defaults to the canvas context on first paint; tests inject an estimate.',
+      factory: function() {
+        var self = this;
+        // Lazy: the canvas context does not exist until the scene is painted once.
+        return function(text, font) {
+          var ctx = self.canvas && self.canvas.context;
+          // TextUtil is a LIB (not a class), so it is referenced directly rather than through requires.
+          if ( ! ctx ) return foam.graphics.TextUtil.estimateMeasurer(7)(text, font);
+          self.measure = foam.graphics.TextUtil.canvasMeasurer(ctx);
```
Comment (abdelaziz-mahdy) [inline src/foam/graphics/Scene.js]: [bug] Every tooltip measurement marks the whole scene dirty.

The demo hands the tooltip `this.scene.measure`, which is the lazy factory closure. That closure
assigns a new canvasMeasurer to self.measure on each call, and each assignment is a property write
that publishes invalidated.

Memoize in a private field instead of writing the property.

File: src/foam/graphics/CView.js [hunk 3]
```diff
@@ -417,6 +417,11 @@ foam.CLASS({
   // Fires when this CView is invalidated and needs a repaint.
   // Is listened to a foam.u2.Canvas() if one was created for
   // this CView.
+  constants: {
+    CACHE_SCALE_TOLERANCE: 0.1,  // re-render the cached bitmap when the effective scale drifts more than 10%
```
Comment (abdelaziz-mahdy) [inline src/foam/graphics/CView.js]: [question] The cache never survives a zoom-in, at any zoom level.

The tolerance is 10% and the demo's zoom step is 1.1, so `|1.1s - s|` comes out just over `0.1s`
and the bitmap is stale on every tick in. Zooming out is 0.0909 and always reuses.

Was 10% picked against the 1.1 step? A tolerance above the step, or re-rendering on a debounce
rather than per tick, would let it hold.

File: src/foam/graphics/CView.js [hunk 4]
```diff
@@ -679,6 +695,93 @@ foam.CLASS({
       }
     },
 
+    function cache() {
+      /**
+       * Like Konva node.cache(): render this subtree once to an offscreen bitmap and
+       * blit it on later paints until something inside changes. Use on a subtree with
+       * many static children (a strip of shapes) so one change elsewhere does not repaint it.
+       * Children added after this call are not watched until cache() is called again.
```
Comment (abdelaziz-mahdy) [inline src/foam/graphics/CView.js]: [question] add() and remove() treat a cached subtree differently.

cache()'s doc says children added later are not watched, so that part is deliberate. But removing a
child does invalidate, by accident — removeChild_ writes c.parent, and c is subscribed. So one half
of the pair silently keeps a stale bitmap and the other does not.

Is the intent that callers re-cache after any structural change? If so, remove() invalidating is
the odd one out.

File: src/foam/animation/Tween.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      return this;
+    },
+
+    function cancel() {
+      this.gen_++;
+      this.running = false;
+    }
+  ],
+
+  static: [
```
Comment (abdelaziz-mahdy) [inline src/foam/animation/Tween.js:71]: [question] Is Tween meant to replace foam/animation, or sit beside it?

Interp.js has linear/easeIn/easeOut/accelerate/oscillate/spline, and Animation.js:96-122 is already
a framed tick with duration, interp, onEnd and an onDetach stop. Neither is in pom.js.

If Tween supersedes them, worth saying so and retiring them; three timing loops is the alternative.

Tween itself has no detach hook, so a tween started by a view that is then detached keeps firing.

File: src/foam/graphics/CView.js [hunk 6]
```diff
@@ -679,6 +695,93 @@ foam.CLASS({
...
+    function cacheBounds_() {
+      /**
+       * Size of the area the bitmap must cover, in local units. A node with its own width/height
+       * (a Box) uses that; a sizeless group (a SceneLayer) uses the extent of its children.
+       */
+      if ( this.width && this.height ) return { w: this.width, h: this.height };
+      var w = 0, h = 0;
+      for ( var i = 0 ; i < this.children.length ; i++ ) {
+        var c = this.children[i], cb = c.cacheBounds_ ? c.cacheBounds_() : { w: c.width, h: c.height };
+        w = Math.max(w, c.x + cb.w * ( c.scaleX || 1 ));
```
Comment (abdelaziz-mahdy) [inline src/foam/graphics/CView.js]: [nice to have] A group child at a negative x falls outside the cached bitmap and does not paint.

The width starts at 0 and only ever takes a max, so a child at x=-50 contributes nothing and the
bitmap is sized from the origin. Demo coordinates are all >= 0, so nothing shows it today.

Track min as well as max, and offset the translate by it.

File: src/foam/graphics/CView.js (hunk 1 above)
Comment (ajeetgill) [reply src/foam/graphics/CView.js]: Confirmed, and the fake context was exactly why the tests missed it. `doTransform` runs before the cache block, so the context scale already includes `scaleX`; both reads multiplied it in again (bitmap = scaleX² the pixels). Dropped the multiplier in `renderCache_` and the stale check (`dad0cf18bc`), and `CViewCacheTest` now uses a fake with `getTransform` so this path runs: a scaleX-2 box under a scale-2 context allocates 216 px wide, not 432.

File: src/foam/graphics/Scene.js (hunk 2 above)
Comment (ajeetgill) [reply src/foam/graphics/Scene.js]: Right on the mechanism: `initCView` routes every propertyChange into `invalidate_`, so the write in the lazy closure dirtied the scene per measurement, and every holder of that closure kept hitting it. For the tooltip itself the extra invalidates were masked because `layout()` writes width/height right after, but a scene that hands `scene.measure` to each element it builds paid one publish per label. Memoized in a private `measurer_` (`4cd2bcd3a5`); the property is never written by the closure now. `SceneTest` asserts zero propertyChange events across two measurements through the held closure.

File: src/foam/graphics/CView.js (hunk 3 above)
Comment (ajeetgill) [reply src/foam/graphics/CView.js]: No, the 10% was not chosen against the step, and your arithmetic is right: 1.1s − s comes out just over 0.1s, so every tick in re-rendered while ticks out reused. Tolerance is 0.25 now and the check is a ratio (`r > 1.25 || r < 1/1.25`), so in and out match: two 1.1 steps blit the bitmap, the third re-renders. I went with a wider threshold rather than a debounce because a debounce needs a timer and a zoom-ended signal CView has no way to get; the constant's comment now says it must stay above one zoom step. Test walks 1 → 1.1 → 1.21 → 1.331 and back.

File: src/foam/graphics/CView.js (hunk 4 above)
Comment (ajeetgill) [reply src/foam/graphics/CView.js]: Confirmed, and the accidental half was the remove side: `removeChild_` writes `c.parent` on a still-subscribed child, `addChild_` writes it on an unsubscribed one. Rather than make remove stop invalidating, I made both structural changes explicit: `addChild_` drops the nearest cached ancestor's bitmap and subscribes the new subtree, `removeChild_` re-caches that ancestor so the removed subtree stops invalidating it (`dad0cf18bc`); `cache()`'s doc says so now. One cost to know about: adding many children one at a time under an already-cached node re-subscribes per add, so build the subtree, then attach. Tests cover both directions, including that a removed child no longer invalidates.

File: src/foam/animation/Tween.js (hunk 5 above)
Comment (ajeetgill) [reply src/foam/animation/Tween.js:71]: Beside, not instead. Animation animates an object's properties by diffing before/after states and interpolating the changed ones; Tween is one scalar over time with an injectable clock, which is what the camera and pulse code wanted and what makes it testable without a global clock stub. To make that literal I've moved it: it's `foam.animation.Tween` now, next to Animation and Interp, with the doc saying which does what (`a9bc5941d4`). Rebuilding Animation on top of Tween so both share one clock is the natural follow-up, in its own PR since it means porting the spacewars and maze demos; I'll open an issue if you agree.

The detach gap was real. `9739c2a822`: `Tween.detach()` cancels, so `owner.onDetach(tween)` stops a tween when its view goes away; TweenTest covers it (one frame runs, owner detaches, nothing more, no onDone).

File: src/foam/graphics/CView.js (hunk 6 above)
Comment (ajeetgill) [reply src/foam/graphics/CView.js]: Done in `dad0cf18bc`: `cacheBounds_` now tracks min as well as max and returns `{ x, y, w, h }`; the offscreen paint translates by `(pad − x, pad − y)` and the blit starts at the extent corner instead of the origin. Test: children at x −50 and x 100 give a 190-wide bitmap that blits from −54 (pad 4).

File: src/foam/graphics/CView.js [hunk 7]
```diff
@@ -760,16 +881,28 @@ foam.CLASS({
...
+      var a = this.cachedAncestor_();
+      if ( a ) { a.invalidateCache(); a.subscribeSubtree_(c); }
       return c;
     },
 
     function removeChild_(c) {
       c.parent = undefined;
       c.canvas = undefined;
+      // Under a cached node: re-cache so the removed subtree stops invalidating it.
+      var a = this.cachedAncestor_();
```
Comment (abdelaziz-mahdy) [inline src/foam/graphics/CView.js:910]: [bug] Removing children one at a time from a cached subtree now rebuilds every subscription on
each removal.

`removeChild_` calls `a.cache()`, which detaches the whole `cacheSubs_` list and re-walks the
subtree; `addChild_` two lines up already does the cheap thing — `invalidateCache()` plus
`subscribeSubtree_(c)` on just that child. `SceneDemo.js:83` removes 600 children singly, so that
is ~360,000 subscribe/detach calls instead of 600.

Mirror the add path: invalidate, and unsubscribe only the removed subtree.

File: src/foam/graphics/CView.js [hunk 8]
```diff
@@ -679,6 +695,107 @@ foam.CLASS({
...
+    },
+
+    function cacheBounds_() {
+      /**
+       * Area the bitmap must cover, in local units: { x, y, w, h }. A node with its own
+       * width/height (a Box) uses that from its origin; a sizeless group (a SceneLayer) uses
+       * the extent of its children, which may start left of or above the origin.
+       */
+      if ( this.width && this.height ) return { x: 0, y: 0, w: this.width, h: this.height };
+      var x0 = 0, y0 = 0, x1 = 0, y1 = 0;
```
Comment (abdelaziz-mahdy) [inline src/foam/graphics/CView.js:766]: [nice to have] A group whose children all sit right of the origin still sizes its bitmap from the
origin.

`x0`/`x1` seed at 0, so the min and max always swallow it. The test's own case — one 40x20 child at
(300,120), `CViewCacheTest.js:88-91` — allocates 348x148 for a 40x20 child.

Seed from the first child when there is one; the blit already handles a non-zero `cacheX_`.

File: src/foam/graphics/CView.js (hunk 7 above)
Comment (ajeetgill) [reply src/foam/graphics/CView.js:910]: Confirmed. `cache()` detaches the whole list and re-walks everything still attached, once per removal. Measured with a spy on the children's `propertyChange.sub`: removing 50 children one at a time re-subscribed 1,225 times, so 600 gives ~180k subscribes plus as many detaches — your number.

Now mirrors the add path. `cacheSubs_` is a `Map` keyed by node (`CView.js:717`), and `unsubscribeSubtree_(node)` detaches only that node and its descendants (`CView.js:721`); `removeChild_` calls `invalidateCache()` plus `unsubscribeSubtree_(c)` (`CView.js:911`). With no argument it still detaches everything, which `cache()` and `uncache()` use.

Tests: a sibling of the removed child keeps its subscription and still invalidates the bitmap (`CViewCacheTest.js:142`), and removing 50 children singly re-subscribes nothing — the assertion prints the count it saw, 1,225 before the fix, 0 after (`CViewCacheTest.js:155`). Commit 3bdcf4be2a.

File: src/foam/graphics/CView.js (hunk 8 above)
Comment (ajeetgill) [reply src/foam/graphics/CView.js:766]: Confirmed — the test even asserted the waste (340 wide for a 40-wide child). The first child now seeds the extent and later children widen it (`CView.js:771`); the origin is no longer part of it. Blit already reads `cacheX_`/`cacheY_`, so nothing else moved. The test now expects `cacheX_ 300, cacheY_ 120, 40+2pad x 20+2pad` (`CViewCacheTest.js:91`).

One caveat added to the doc comment: a sizeless node that paints something itself is not measured either way; it should carry a width/height. Commit 3bdcf4be2a.
